apps/aefate/src/aefa_fate.erl

```diff
@@ -306,60 +315,118 @@ bind_args_from_signature({ArgTypes, _RetSignature}, ES) ->
-check_type(boolean, B) when ?IS_FATE_BOOLEAN(B) -> true;
-check_type(string, S) when ?IS_FATE_STRING(S) -> true;
-check_type(address, A) when ?IS_FATE_ADDRESS(A) -> true;
-check_type(contract, A) when ?IS_FATE_CONTRACT(A) -> true;
-check_type(oracle, A) when ?IS_FATE_ORACLE(A) -> true;
-check_type(name, A) when ?IS_FATE_NAME(A) -> true;
-check_type(bits, B) when ?IS_FATE_BITS(B) -> true;
-check_type({list, any}, L) when ?IS_FATE_LIST(L) ->
-    true;
-check_type({list, ET}, L) when ?IS_FATE_LIST(L) ->
-    check_same_type(ET, ?FATE_LIST_VALUE(L));
-check_type({tuple, Elements}, T) when ?IS_FATE_TUPLE(T) ->
-    check_all_types(Elements, ?FATE_TUPLE_ELEMENTS(T));
-check_type({map, Key, Value}, M) when ?IS_FATE_MAP(M) ->
-    {Ks, Vs} = lists:unzip(maps:to_list(?FATE_MAP_VALUE(M))),
-    check_same_type(Key, Ks) andalso
-    check_same_type(Value, Vs);
-check_type({variant, Types}, V) when ?IS_FATE_VARIANT(V) ->
-    ?FATE_VARIANT(Arities, Tag, Value) = V,
-    check_variant_arities(Types, Arities) andalso
-        check_type(lists:nth(Tag + 1, Types), ?FATE_TUPLE(Value));
-check_type(_T, _V) -> false.
-
-check_variant_arities([?FATE_TUPLE(Types)|Left1], [Arity|Left2]) ->
-    Arity =:= length(Types) andalso check_variant_arities(Left1, Left2);
-check_variant_arities([], []) ->
-    true;
-check_variant_arities(_, _) ->
-    false.
+check_type(T, V) ->
+    try
+        match_type(T, infer_type(V))
+    catch throw:_Err ->
+        false
+    end.
+
+infer_type(X) when ?IS_FATE_INTEGER(X)   -> integer;
+infer_type(X) when ?IS_FATE_BOOLEAN(X)   -> boolean;
+infer_type(X) when ?IS_FATE_STRING(X)    -> string;
+infer_type(X) when ?IS_FATE_ADDRESS(X)   -> address;
+infer_type(X) when ?IS_FATE_HASH(X)      -> hash;
+infer_type(X) when ?IS_FATE_SIGNATURE(X) -> signature;
+infer_type(X) when ?IS_FATE_CONTRACT(X)  -> contract;
+infer_type(X) when ?IS_FATE_ORACLE(X)    -> oracle;
+infer_type(X) when ?IS_FATE_NAME(X)      -> name;
+infer_type(X) when ?IS_FATE_BITS(X)      -> bits;
+infer_type(X) when ?IS_FATE_LIST(X) ->
+    {list, infer_element_type(?FATE_LIST_VALUE(X))};
+infer_type(X) when ?IS_FATE_MAP(X) ->
+    {Ks, Vs} = lists:unzip(maps:to_list(?FATE_MAP_VALUE(X))),
+    KeyT = infer_element_type(Ks),
+    ValT = infer_element_type(Vs),
+    {map, KeyT, ValT};
+infer_type(X) when ?IS_FATE_TUPLE(X) ->
+    {tuple, [infer_type(Y) || Y <- ?FATE_TUPLE_ELEMENTS(X)]};
+infer_type(X) when ?IS_FATE_VARIANT(X) ->
+    ?FATE_VARIANT(Arities, Tag, Value) = X,
+    N    = length(Arities),
+    [ throw({bad_variant_tag, Tag, Arities}) || Tag >= N ],
+    Type = infer_type(?FATE_TUPLE(Value)),
+    %% Partial variant type (only arities for other constructors)
+    {variant, [ if I == Tag -> Type; true -> A end
+                || {A, I} <- lists:zip(Arities, lists:seq(0, N - 1))]};
+infer_type(X) -> throw({not_a_fate_value, X}).
+
+infer_element_type(Xs) ->
+    lists:foldl(fun intersect_types/2, any, [infer_type(X) || X <- Xs]).
```

Comment:
I understand intersect, and not union, because union type is expressed as variant type, and value ov variant type needs to be expressed [as constructor](https://github.com/aeternity/protocol/blame/34e8eeee3c9d3c7fafa83883bc7d95273315125d/contracts/sophia.md#L1282-L1287).